data_structures/skip_list.cpp

```diff
@@ -0,0 +1,212 @@
+/**
+ * A skip list is a data structure that is used for storing a sorted list of items with a
+ *  help of hierarchy of linked lists that connect increasingly sparse subsequences of the items
+ * 
+ *  References used: GeeksForGeeks skip list code, https://iq.opengenus.org/skip-list/ Pseudo Code.
+*/
+
+#include <iostream>
+#include <vector>
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+#include <time.h>   
+#include <cstdlib>
+
+// using namespace std; 
+using std::vector;
+using std::endl;
+
+#define MAXLVL 2
+#define P 0.5
+  
+
+/**
+ *  Node structure [Key][Node*, Node*...]
+*/
+class Node { 
+public: 
+    int key; 
+    void* value;
+    /*Forward Array*/
+    vector<Node*> forward;
+    Node(int key, int level, void* value); 
+};
+
+  
+Node::Node(int key, int level, void* value) { 
+    this->key = key; 
+  
+    /*Initialization of forward vector*/
+    for (int i = 0; i < sizeof(Node*)*(level+1); i++){
+        forward.push_back(NULL);
+    }
+} 
+  
+// Class for Skip list 
+class SkipList { 
+    int level; 
+    Node *header; 
+public: 
+    SkipList(); 
+    int randomLevel(); 
+    void insertElement(int, void*); 
+    void deleteElement(int); 
+    void* searchElement(int); 
+    void displayList(); 
+    ~SkipList();
+};
+
+
+  
+SkipList::SkipList() { 
+    level = 0; 
+    /* Header initialization*/
+    header = new Node(-1, MAXLVL, NULL); 
+} 
+
+
+SkipList::~SkipList(){
+    delete header;
+    for(int i=0; i <= level; i++) { 
+        Node *node = header->forward[i]; 
+        Node* temp;
+        while(node != NULL) { 
+            temp = node;
+            node = node->forward[i];
+            delete temp;
+        } 
+    } 
+}
+
+  
+/**
+ * Returns random level for skip list;
+*/
+int SkipList::randomLevel() { 
+    int lvl = 0; 
+    while(static_cast<float>(std::rand())/RAND_MAX < P && lvl < MAXLVL) lvl++;  
+    return lvl; 
+} 
+  
+  
+// Insert given key in skip list 
+void SkipList::insertElement(int key, void* value) { 
+    std::cout << "Inserting" << key << "...";
+    Node *x = header; 
+    Node *update[MAXLVL+1]; 
+    memset(update, 0, sizeof(Node*)*(MAXLVL+1)); 
+
+
+    for(int i = level; i >= 0; i--) { 
+        while(x->forward[i] != NULL && x->forward[i]->key < key)  x = x->forward[i]; 
+        update[i] = x; 
```

Comment:
this is invalid because only `update[0]` is defined.

Reply:
I don't get it. size of update is (MAXLVL+1)* sizeof(Node*); on 95 line memset is used for initialization.

Reply:
`memset` does not create, it simply sets the values. If out of bounds, segmentation fault will occur as shown in the screenshot. 
please ensure the code compiles and executes correctly. All reviewers and contributors are volunteers devoting their own time. Hence, a self-assessment of the code by the authors will be helpful.

Reply:
Yes, I understand and I'm very glad for devotion. memset shouldn't happen to reach out of bounds as size of update array equals  (MAXLVL+1)* sizeof(Node*); which is also bounds for memset.
I devoted hours studying the code yesterday and I find it very elegant implementation.

Reply:
understood - `update` is an array of pointers.